lib/galaxy/tools/__init__.py

```diff
@@ -2956,7 +2956,13 @@ def produce_outputs(self, trans, out_data, output_collections, incoming, history
 
         for i, incoming_repeat in enumerate(incoming["datasets"]):
             if incoming_repeat["input"]:
-                new_elements["%d" % i] = incoming_repeat["input"].copy(copy_tags=incoming_repeat["input"].tags)
+                if incoming_repeat["id_cond"]["id_select"] == 'idx':
+                    identifier = str(i)
+                elif incoming_repeat["id_cond"]["id_select"] == 'identifier':
+                    identifier = incoming_repeat["input"].name
```

Comment:
Can you check if we have access to the element identifier here ? You can map this tool over a collection, so I think the element identifier would be the right attribute to use if it is there.

Reply:
Well, I guess that's a no:
```
galaxy.tools ERROR 2021-09-10 13:02:30,579 [pN:main,p:2140,tN:ThreadPoolExecutor-0_8] Exception caught while attempting to execute tool with id '__BUILD_LIST__':
Traceback (most recent call last):
  File "/home/runner/work/galaxy/galaxy/galaxy root/lib/galaxy/tools/__init__.py", line 1750, in handle_single_execution
    flush_job=flush_job,
  File "/home/runner/work/galaxy/galaxy/galaxy root/lib/galaxy/tools/__init__.py", line 1838, in execute
    return self.tool_action.execute(self, trans, incoming=incoming, set_output_hid=set_output_hid, history=history, **kwargs)
  File "/home/runner/work/galaxy/galaxy/galaxy root/lib/galaxy/tools/actions/model_operations.py", line 61, in execute
    self._produce_outputs(trans, tool, out_data, output_collections, incoming=incoming, history=history, tags=preserved_tags, hdca_tags=preserved_hdca_tags)
  File "/home/runner/work/galaxy/galaxy/galaxy root/lib/galaxy/tools/actions/model_operations.py", line 77, in _produce_outputs
    tool.produce_outputs(trans, out_data, output_collections, incoming, history=history, tags=tags, hdca_tags=hdca_tags, tag_handler=tag_handler)
  File "/home/runner/work/galaxy/galaxy/galaxy root/lib/galaxy/tools/__init__.py", line 2962, in produce_outputs
    identifier = incoming_repeat["input"].element_identifier
AttributeError: 'HistoryDatasetAssociation' object has no attribute 'element_identifier'
```

Reply:
We can try/except .. but I guess when mapped over the tool will still get an HDA, or?

Reply:
Yes, but at that point we may have set the element_identifier ... I think that maybe be happening in https://github.com/jmchilton/galaxy/blob/6412363dcca1c005cd93d48d5c72354d46877b45/lib/galaxy/tools/parameters/meta.py#L252

Reply:
Also I'd suggest `getattr(incoming_repeat["input"], 'element_identifier', incoming_repeat["input"].name)` ... but only if the mapping over works :)

Reply:
According to the nice figures that have been added recently to the help (https://github.com/galaxyproject/galaxy/pull/11068) mapping over should work. But I'm not really sure how to test (manually) if the element identifier is really used. Do you have an idea?

Reply:
I meant whether the element identifiers are really being added at this point. Yes, mapping over definitely works (and #11068 took some shortcuts to explain things that are more complex ...)

Reply:
You'd have to add an API test if you want to be sure, there's no way to test this with the tool testing framework.

Reply:
> You'd have to add an API test if you want to be sure, there's no way to test this with the tool testing framework.

I just wanted to do a manual test and was wondering how I can produce a difference in name and identifier that I can observe in the output?

Reply:
You can create a list in the UI and rename some of the items.
<img width="765" alt="Screenshot 2021-09-15 at 18 10 37" src="https://user-images.githubusercontent.com/6804901/133469840-a533f09a-f8db-4c4f-92a2-b02c57e5336a.png">